### ISSUE TITLE
Ensure Ubuntu packages for upcoming releases (LTS or otherwise) become available on a timely manner

### ISSUE THREAD

**Issue by behrangsa** [docker/for-linux#940]
* [x] This is a feature request

### Expected behavior

This is in relation to #832 (and similar issues before it). 

Instead of waiting for users to raise a ticket once every six months, and spending another 2-4 months to resolve the ticket, ensure Ubuntu packages are ready on a timely manner, ideally on day 1.

Ubuntu has a very well defined release cycle (once every six months, on April and October of each year -- [see here](https://ubuntu.com/about/release-cycle)). Before the final version is released, Ubuntu provides a release candidate too.

In this day and age, there's no excuse for releasing Docker for an OS with such a well defined release cycle very late all the time.

### Actual behavior

Docker uploads packages for the latest versions of Ubuntu very late. Often, two months after Docker makes the packages available, the next version of Ubuntu is already released.

See #832.

**Comment by DeeDeeG** [comment]
re: @graingert's comment here: https://github.com/docker/for-linux/issues/832#issuecomment-595139191 and the discussions in #832 (and for prior Ubuntu releases this has also come up), well, there is no beta available yet. That happens consistently about a month before release. But there are daily ISO's for testing. (These have been available since a few days after 19.10 was released, i.e. about four months ago.)

For an example, please everyone who is curious about the Ubuntu release schedule, review the schedule for Ubuntu 20.04 "Focal Fossa": https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule

But I must emphasize a few things:

- This schedule is essentially the same every time. Count roughly the same number of weeks back from any scheduled final release, and the same milestones will come up.
  - A lot of stuff is frozen and stabilized in the final two months before a release.
- The final release date for the upcoming release is published almost as soon as the prior release is, well, released.
- **Daily ISO's** for an upcoming release are published _**a day or two after**_ the prior release is, er, released. They are popular for testing, and many people run them as daily driver OS's, including myself.

As such, it is indeed very reliable and predictable to reason about (and test) upcoming releases. I hope that with this information, it becomes clear that targeting upcoming releases is very achievable. The full information about an upcoming release is available basically a half a year (six months) in advance.

---

Just for rhetorical purposes and to make a point: The general timing of an upcoming release could also be predicted a decade in advance easily, as the schedule for beta/final/etc. is the same every release, and final release is always "late April" then "late October" every year.

(And LTS happens to be every even-numbered year, in late April.)

I can say right now there will be a Ubuntu 30.04 released in late April 2030. Beta will be in early April or late March 2030. Daily ISOs will be available in late October 2029. That's how reliable and consistent Ubuntu releases are, again just to prove the point.

I don't mean to call anyone out, since I am appreciative of this open-source project offered at no cost. I only want to make this information available in case it actually helps.

**Comment by darkdragon-001** [comment]
As shown in the [20.04 Focal Fossa Release Schedule](https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule), the new toolchain is uploaded one week after releasing the previous version.

What about creating a configuration right away. This way, the next Docker release will be uploaded with the CI then, if I understood the process correctly. Further, you will notice problems right away and can fix them early.

**Comment by graingert** [comment]
ubuntu 20.04 has been released and:
```
E: The repository 'https://download.docker.com/linux/ubuntu focal Release' does not have a Release file.
```

**Comment by darkdragon-001** [comment]
The toolchain for Ubuntu 20.10 _Groovy Gorilla_ [will be ready this week](https://discourse.ubuntu.com/t/groovy-gorilla-release-schedule/15531). So when you add it to your CI directly, maybe it will be on time for the next release :blink:

**Comment by alfredbez** [comment]
I ran into this issue today and installed docker via `docker.io`, worked perfectly for me.

I used these guides:
* https://linuxconfig.org/how-to-install-docker-on-ubuntu-20-04-lts-focal-fossa
* https://linuxconfig.org/how-to-install-docker-compose-on-ubuntu-20-04-focal-fossa-linux

Why is this way not mentioned in the [install page](https://docs.docker.com/engine/install/ubuntu/)? Will I run into any problems by using this way?

**Comment by SuperSandro2000** [comment]
docker.io is the debian docker package and has nothing to do with this github repo.

**Comment by SuperSandro2000** [comment]
> ubuntu 20.04 has been released and:
> 
> ```
> E: The repository 'https://download.docker.com/linux/ubuntu focal Release' does not have a Release file.
> ```

just use eoan for now.

**Comment by DeeDeeG** [comment]
For what it's worth, Ubuntu's version of docker.io only _used to be_ woefully out of date.

Now it's fairly recent. They have also been updating their version of docker.io throughout the lifecycle of their distros, long after release. I think any way you can get it up and running, practically speaking, is going to be about equivalent. For the moment, both Ubuntu and the Docker company are on version 19.3.x

The difference is, if you have an issue with using the Ubuntu docker.io package, you can (and should) go to Ubuntu first for support. If you use the version direct from the Docker company, then bug reports/support requests can go here. (Or in either case, ask the community, like on stackexchange or AskUbuntu or social media.)

My two cents as a Docker user (I'm not part of the Docker org.)

**Comment by AlistairB** [comment]
This is not working for me.

```bash
$ sudo add-apt-repository \
   "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
   $(lsb_release -cs) \
   stable"
# works fine

$ sudo apt update
# works fine

$ sudo apt install -y \
  docker-ce \
  docker-ce-cli \
  containerd.io

Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package docker-ce is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

Package docker-ce-cli is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'docker-ce' has no installation candidate
E: Package 'docker-ce-cli' has no installation candidate
```

I tried with a local ubuntu package server and with `Main Server`.

**Comment by StefanScherer** [comment]
Docker 19.03.9 is released, with packages for Ubuntu Focal Fossa.
Thanks for you patience. Enjoy!

**Comment by graingert** [comment]
@StefanScherer do you know when packages for 20.10 will be available?

**Comment by Barbiero** [comment]
I guess it's that time of the year again

Any sign of support for 20.10, which was released today? ;)

**Comment by graingert** [comment]
What preparations are being made to prevent this delay for 21.04?

**Comment by memeplex** [comment]
Is it possible to run packages por 20.04 in 20.10? docker is the only dependency keeping me from upgrading.

**Comment by DeeDeeG** [comment]
@memeplex yes, that usually works quite well, in my experience.

Your options for upgrading Ubuntu right away and installing the latest Docker (that I am aware of, in no particular order) are:
- Use the packages from Docker for the previous Ubuntu release
- Use the packages from the Ubuntu repositories
  - you can get `docker.io` and `containerd` (and also `docker-compose` if you want) from the Ubuntu repositories. (These are named differently from the `docker-ce`/`docker-ce-cli` and `containerd.io` package names from Docker, but they are mostly the same thing).
  - You're not supposed to mix and match the packages from Ubuntu with the packages from Docker. I don't think `apt` will let you, but try to be aware of which packages you already have installed from what source before switching over, and maybe just uninstall all of them before switching over, just to be thorough.)
- If you're an intrepid software builder, there is an open-source repository that can build the Docker packages against various Ubuntu/Debian/Fedora (and so on) distributions.
  - Since this builds from the `master` branch of the Docker/containerd source repos by default, rather than a release milestone, I don't think this would get any real support if you run into problems. Anyway, here's those repositories:
    - https://github.com/docker/docker-ce-packaging
    - https://github.com/docker/containerd-packaging

**Comment by memeplex** [comment]
Thank you very much for the detailed info. `docker.io` seems to be discouraged since it's not ever mentioned in the official documentation. Nevertheless it might be the case that it's not that outdated nowadays (see https://askubuntu.com/questions/1182820/why-is-docker-not-in-official-ubuntu-repos/1182826#1182826). What's is your recommendation or experience regarding it?

**Comment by DeeDeeG** [comment]
@memeplex anecdotally, I can't speak for anyone but myself, but yes, the `docker.io` package from Ubuntu is fine. And yes, it's more up-to-date now than it used to be. Ubuntu Groovy ships [the same version number of Docker (19.03.13)](https://packages.ubuntu.com/groovy/docker.io) that is available [in the Docker repos](https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/amd64/).

I personally use the package from Ubuntu (the `docker.io` package) lately, because it's simply an easier way to get Docker. `apt-get install docker.io` and you're done installing Docker.

Ubuntu doesn't seem to modify the package very much for their repos, and it's kept fairly up-to-date. Regardless of where you get it from, I think they are equivalent. To be clear, I have nothing against the official package in Docker's repositories. Whichever you prefer, I don't think there is a noticeable difference for 99.99% of people. My two cents. I say get it from wherever you prefer to get it from.

**Comment by DeeDeeG** [comment]
Update: I just noticed Docker for Ubuntu 20.10 "Groovy Gorilla" is available in [the `test` section of Docker's repositories](https://download.docker.com/linux/ubuntu/dists/groovy/pool/test/amd64/).

(`containerd.io` is available in [the `stable` section of Docker's repositories for Groovy](https://download.docker.com/linux/ubuntu/dists/groovy/pool/stable/amd64/), whereas `docker-ce` and `docker-ce-cli` are not in `stable` yet; They are only in `test` for the moment.)

(I have no idea when `docker-ce` and `docker-ce-cli` will be in the `stable` repository for Groovy, but hopefully soon? 🤷 Time will tell.)

---

Instructions to install the packages from the `test` repository (for those who can responsibly handle the risks of running pre-release software) are available here: https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository

Basically you need to enable the `test` repo in addition to the `stable` one in your APT sources.

> To add the **nightly** or **test** repository, add the word `nightly` or `test` (or both) after the word `stable` in the commands below.

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as release cadence should be more closely to distro releases now.
